Azure cloudprovider retry using flowcontrol

jackfrancis opened the pull request:
An initial attempt at engaging exponential backoff for API error responses.

Addresses #47048

Uses k8s.io/client-go/util/flowcontrol; implementation inspired by GCE
cloudprovider backoff.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

The existing azure cloudprovider implementation has no guard rails in place to adapt to unexpected underlying operational conditions (i.e., clogs in resource plumbing between k8s runtime and the cloud API). The purpose of these changes is to support exponential backoff wrapping around API calls; and to support targeted rate limiting. Both of these options are configurable via `--cloud-config`.

Implementation inspired by the GCE's use of `k8s.io/client-go/util/flowcontrol` and `k8s.io/apimachinery/pkg/util/wait`, this PR likewise uses `flowcontrol` for rate limiting; and `wait` to thinly wrap backoff retry attempts to the API.

**Special notes for your reviewer**:


Pay especial note to the declaration of retry-able conditions from an unsuccessful HTTP request:
- all `4xx` and `5xx` HTTP responses
- non-nil error responses

And the declaration of retry success conditions:
- `2xx` HTTP responses

Tests updated to include additions to `Config`.

Those may be incomplete, or in other ways non-representative.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Added exponential backoff to Azure cloudprovider
```

jackfrancis committed:
Azure cloudprovider retry using flowcontrol

An initial attempt at engaging exponential backoff for API error responses.

Uses k8s.io/client-go/util/flowcontrol; implementation inspired by GCE
cloudprovider backoff.

jsafrane commented:
I know very little about azure API behavior.

/unassign
/assign @colemickens

jsafrane commented:
@k8s-bot ok to test

brendandburns commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
nit: 2017

brendandburns commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,149 @@
+	if err != nil {
+		return true
+	}
+	if resp.StatusCode == 429 || resp.StatusCode == 500 {
```
Should this be 5xx instead of just 500?

Also, please use http.StatusServerError etc. rather than hard constants.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,149 @@
+// isSuccessHTTPResponse determines if the response from an HTTP request suggests success
+func isSuccessHTTPResponse(resp autorest.Response) bool {
+	// TODO determine the complete set of success conditions
+	if resp.StatusCode == 200 || resp.StatusCode == 201 || resp.StatusCode == 202 {
```
use `http.StatusOK` etc instead of constants.

brendandburns commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -177,6 +179,9 @@ func NewCloud(configReader io.Reader) (cloudprovider.Interface, error) {
 	az.StorageAccountClient.Authorizer = servicePrincipalToken
 
+	// 1 qps, up to 5 burst when in flowcontrol; i.e., aggressive backoff enforcement
+	az.operationPollRateLimiter = flowcontrol.NewTokenBucketRateLimiter(1, 5)
```
You don't ever use this as far as I can tell.

I think we really want this to be used.

brendandburns replied:
Let's make this configurable (and disable-able) via config file.

jackfrancis replied:
@brendandburns  @colemickens would the `--cloud-config` --> `Config` struct be the best configuration vector for this?

brendandburns replied:
yes.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,149 @@
+
+// CreateOrUpdateSGWithRetry invokes az.SecurityGroupsClient.CreateOrUpdate with exponential backoff retry
+func (az *Cloud) CreateOrUpdateSGWithRetry(sg network.SecurityGroup) error {
+	return wait.Poll(operationPollInterval, operationPollTimeoutDuration, func() (bool, error) {
```
How is this exponential backoff? Poll always just waits for Interval

jackfrancis committed:
errata, wait.ExponentialBackoff, regex HTTP codes

- corrected Copyright copy/paste
- now actually implementing exponential backoff instead of regular interval retries
- using more general HTTP response code success/failure determination (e.g., 5xx for retry)
- net/http constants ftw

jackfrancis commented:
@brendandburns thanks for keeping me honest, review notes incorporated

jackfrancis committed:
errata

arg cruft in CreateOrUpdateSGWithRetry function declaration

brendandburns commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,185 @@
+		return true
+	}
+	// HTTP 5xx suggests we should retry
+	r, err := regexp.Compile(`^5\d\d$`)
```
Use numeric comparisons

eg

```
code >= http.StatusInternalError
```

Do this for the 2xx below too.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,185 @@
+// isSuccessHTTPResponse determines if the response from an HTTP request suggests success
+func isSuccessHTTPResponse(resp autorest.Response) bool {
+	// HTTP 2xx suggests a successful response
+	r, err := regexp.Compile(`^2\d\d$`)
```
As above

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -149,7 +149,13 @@ func (az *Cloud) EnsureLoadBalancer(clusterName string, service *v1.Service, nod
+		if shouldRetryAPIRequest(resp, err) {
+			retryErr := az.CreateOrUpdateSGWithRetry(sg)
+			if retryErr != nil {
+				return nil, retryErr
```
Do we want to return here? Or just set `err = retryErr` and let the code below handle both err cases?

I think I prefer that approach.

jackfrancis replied:
Yeah, your approach is better. The purpose of the novel `retryErr` reference (as opposed to reusing the existing `err` reference prior to checking for retry-ability is to be able to facilitate logging/debug in the retry execution branch. We can still do that by reusing the `err` reference and eliminate an unnecessary `return`.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -310,7 +322,13 @@ func (az *Cloud) EnsureLoadBalancerDeleted(clusterName string, service *v1.Servi
+			resp, err := az.SecurityGroupsClient.CreateOrUpdate(az.ResourceGroup, *reconciledSg.Name, reconciledSg, nil)
+			if shouldRetryAPIRequest(resp, err) {
+				retryErr := az.CreateOrUpdateSGWithRetry(reconciledSg)
+				if retryErr != nil {
```
As above

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -339,14 +357,26 @@ func (az *Cloud) cleanupLoadBalancer(clusterName string, service *v1.Service, is
-				_, err = az.LoadBalancerClient.CreateOrUpdate(az.ResourceGroup, *lb.Name, lb, nil)
+				resp, err := az.LoadBalancerClient.CreateOrUpdate(az.ResourceGroup, *lb.Name, lb, nil)
+				if shouldRetryAPIRequest(resp, err) {
+					retryErr := az.CreateOrUpdateLBWithRetry(lb)
```
As above

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -339,14 +357,26 @@ func (az *Cloud) cleanupLoadBalancer(clusterName string, service *v1.Service, is
+				resp, err := az.LoadBalancerClient.Delete(az.ResourceGroup, lbName, nil)
+				if shouldRetryAPIRequest(resp, err) {
+					retryErr := az.DeleteLBWithRetry(lbName)
+					if retryErr != nil {
```
Here too

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -392,7 +422,13 @@ func (az *Cloud) ensurePublicIPExists(serviceName, pipName string) (*network.Pub
+	resp, err := az.PublicIPAddressesClient.CreateOrUpdate(az.ResourceGroup, *pip.Name, pip, nil)
+	if shouldRetryAPIRequest(resp, err) {
+		retryErr := az.CreateOrUpdatePIPWithRetry(pip)
+		if retryErr != nil {
```
And here

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -848,7 +887,13 @@ func (az *Cloud) ensureHostInPool(serviceName string, nodeName types.NodeName, b
+		resp, err := az.InterfacesClient.CreateOrUpdate(az.ResourceGroup, *nic.Name, nic, nil)
+		if shouldRetryAPIRequest(resp, err) {
+			retryErr := az.CreateOrUpdateInterfaceWithRetry(nic)
+			if retryErr != nil {
```
You get the point, I'm going to stop commenting on all of them

jackfrancis replied:
I was enjoying how many variations you could come up with. :)

jackfrancis committed:
two optimizations

- removed unnecessary return statements
- optimized HTTP response code evaluations as numeric comparisons

fejta commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this
ref: kubernetes/test-infra#2932

brendandburns commented:
Looks like you need to run `./hack/update-bazel.sh` otherwise, LGTM.

/lgtm
/approve

jackfrancis committed:
update-bazel.sh mods

jackfrancis committed:
backoff logging, error handling, wait.ConditionFunc

- added info and error logs for appropriate backoff conditions/states
- rationalized log idioms across all resource requests that are backoff-enabled
- processRetryResponse as a wait.ConditionFunc needs to supress errors if it wants the caller to continue backing off

grodrigues3 commented:
@jackfrancis can you reference the issue that this PR addresses (it may be in this repo or features or another repo)?

seanknox commented:
@grodrigues3 here is the issue the ACS team is using to track this work: https://github.com/Azure/acs-engine/issues/635

jdumars commented:
@grodrigues3 I had asked this question during the leadership summit, but didn't get an answer.  Do we need to create a separate issue in k/k that references our issue in ACS?

brendandburns commented:
@jackfrancis thanks for looking into this, it would be great to get this merged asap.

@grodrigues3 Do you us to file a specific kubernetes issue for this, or is the referenced issue sufficient?

Thanks

jackfrancis committed:
configurable backoff

- leveraging Config struct (—cloud-config) to store backoff and rate limit on/off and performance configuration
- added add’l error logging
- enabled backoff for vm GET requests

jackfrancis commented:
@brendandburns see `Config` updates in latest commit.

Both backoff and rate limiting can be configured therefrom, as well as enabled/disabled entirely.

wojtek-t commented:
@jackfrancis - this PR is marked as 1.7 milestone. However, according to the new policy, all non-trivial PRs has to link some issue, so if you want this in 1.7, please attach this PR to some issue.

jdumars commented:
/retest 

Also, I'm going to open an issue for this one referencing the ACS issue.

jdumars commented:
@wojtek-t @grodrigues3 Please see issue [47048](https://github.com/kubernetes/kubernetes/issues/47048)

jackfrancis committed:
test coverage

We want to ensure that backoff and rate limit configuration is opt-in

jackfrancis committed:
rate limiting on all azure sdk GET requests

jackfrancis committed:
az.getVirtualMachine already rate-limited

we don’t need to rate limit the calls _to_ it

brendandburns commented:
@jackfrancis this looks like it has some `go vet` errors.

You can run `go vet k8s.io/kubernetes/pkg/cloudprovider/providers/azure` to generate the errors.

jackfrancis committed:
rate limiting everywhere

not waiting to rate limit until we get an error response from the API, doing so on initial request for all API requests

jackfrancis committed:
go vet errata

jackfrancis commented:
@brendandburns thx, addressed

jackfrancis committed:
preferring float32 for rate limit QPS param

jdumars commented:
/retest

jdumars commented:
/sig-azure

jdumars commented:
/sig azure

jdumars commented:
@grodrigues3 @wojtek-t this should have the 1.7 milestone attached

jdumars commented:
@brendandburns LGTM needed

brendandburns commented:
Will look soon.

--brendan

________________________________
From: Jaice Singer DuMars <notifications@github.com>
Sent: Wednesday, June 7, 2017 8:17:41 AM
To: kubernetes/kubernetes
Cc: Brendan Burns; Mention
Subject: Re: [kubernetes/kubernetes] Azure cloudprovider retry using flowcontrol (#46660)


@brendandburns<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbrendandburns&data=02%7C01%7Cbburns%40microsoft.com%7Cf8da174c34094b61590e08d4adb858d9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636324454654072622&sdata=8FMGW5GCDO49QqNDjYqu%2B0PeyCP99pW4aEj1%2BN6NNdg%3D&reserved=0> LGTM needed

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkubernetes%2Fkubernetes%2Fpull%2F46660%23issuecomment-306827481&data=02%7C01%7Cbburns%40microsoft.com%7Cf8da174c34094b61590e08d4adb858d9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636324454654072622&sdata=hRjGv%2FLJiS2o0vpYCscbhOrQT8CucPYj0mvMTycLcjY%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFfDglQ3yFCPhZqDIOYAyORASDTfw5fcks5sBr8VgaJpZM4NrAPC&data=02%7C01%7Cbburns%40microsoft.com%7Cf8da174c34094b61590e08d4adb858d9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636324454654072622&sdata=XdJV8LkvG%2B6COOD7dPE1SS6pCF63y9JuZ5MwmZ4O8MQ%3D&reserved=0>.

brendandburns commented:
/lgtm
/approve

brendandburns commented:
/approve #47048

brendandburns commented:
/approve

yangl900 commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,170 @@
+		return true
+	}
+	// HTTP 4xx or 5xx suggests we should retry
+	if 399 < resp.StatusCode && resp.StatusCode < 600 {
```
400 is not a retry-able error in Azure, I think should only retry status code == 429 or status code > 500

jackfrancis replied:
@brendandburns I'm happy to incorporate this feedback if you're willing to go through the lgtm approval obstacle course all over agin. @yangl900 the other two remarks I believe we can justify tackling later on as part of a more holistic effort to pair k8s cloudprovider code with specific API idioms

yangl900 commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -32,10 +33,20 @@ import (
+const (
+	// CloudProviderName is the value used for the --cloud-provider flag
+	CloudProviderName      = "azure"
+	rateLimitQPSDefault    = 1.0
```
read quota is much higher than write, I don't know if you want to differentiate in this change. that potentially give you higher quota. we can improve this later too.

yangl900 commented on pkg/cloudprovider/providers/azure/azure_backoff.go:
```diff
@@ -0,0 +1,170 @@
+		var retryErr error
+		machine, exists, retryErr = az.getVirtualMachine(name)
+		if retryErr != nil {
+			glog.Errorf("backoff: failure, will retry,err=%v", retryErr)
```
there is a Retry-After header returned from 429 requests, probably helpful if we trace that.

yangl900 reviewed (commented):
Looks good, just a few comments.

brendandburns commented:
@jackfrancis can you address @yangl900's comments in a follow on PR?

Thanks!
